### PR TITLE
 Avoid using array pool when only one completion callback is registered 

### PR DESCRIPTION
**Pull request by pakrym**
Fixes https://github.com/dotnet/corefx/issues/27938

### CODE REVIEW

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs [hunk 1]
```diff
@@ -40,32 +41,46 @@ public PipeCompletionCallbacks TryComplete(Exception exception = null)
...
+            {
+                _firstCallback = new PipeCompletionCallback(callback, state);
+                _callbackCount++;
+            }
+            else
+            {
+                EnsureSpace();
+
+                // -1 to adjust for _firstCallback
+                var callbackIndex = _callbackCount - 1;
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs]: nit: don't use var.

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs [hunk 2]
```diff
@@ -40,32 +41,46 @@ public PipeCompletionCallbacks TryComplete(Exception exception = null)
...
+                _callbackCount++;
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs]: nit: increment after the declaration/assignment succeeds.

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs [hunk 3]
```diff
@@ -40,32 +41,46 @@ public PipeCompletionCallbacks TryComplete(Exception exception = null)
...
             {
                 _callbacks = s_completionCallbackPool.Rent(InitialCallbacksSize);
             }
 
-            int newIndex = _callbackCount;
-            _callbackCount++;
+            int newLength = _callbackCount - 1;
 
-            if (newIndex == _callbacks.Length)
+            if (newLength == _callbacks.Length)
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs]: `s_completionCallbackPool.Rent(_callbacks.Length * 2);` has potential for integer over flow. How likely is it to hit the overflow (i.e. callbacks.Length > 1 GB)? Check and throw?

This way you also avoid the risk of overflowing when doing `_callbackCount++;`  above.

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletionCallbacks.cs [hunk 4]
```diff
@@ -12,52 +12,65 @@ internal sealed class PipeCompletionCallbacks
...
-                    }
-                    catch (Exception ex)
+                    for (var i = 0; i < _count - 1; i++)
                     {
-                        if (exceptions == null)
-                        {
-                            exceptions = new List<Exception>();
-                        }
-                        exceptions.Add(ex);
+                        var callback = _callbacks[i];
```
Comment (ahsonkhan) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletionCallbacks.cs]: nit: fix use of var here/elsewhere

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs [hunk 5]
```diff
@@ -19,6 +19,7 @@ internal struct PipeCompletion
         private bool _isCompleted;
         private ExceptionDispatchInfo _exceptionInfo;
 
+        private PipeCompletionCallback _firstCallback;
```
[the remark this answers is not in the document]
Comment (halter73) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs]: I'm not sure what you're getting at. What starting length?

Before this change, PipeCompletion would lease a 1-element _callbacks array from the ArrayPool on the first (and almost always only) AddCallback invocation. Of course, the ArrayPool was free to return a larger array. The code was simpler than it is now, but resulted in more arrays being leased.

Are you suggesting just allocating an array?

[the remark this answers is not in the document]
Comment (ahsonkhan) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs]: Never mind. I was thinking something like allocating a static single-element array instead, but we would still need an instance field (which is the `PipeCompletionCallback[]` that grows), and hence it changes nothing. We can't grow the static array for every instance of `PipeCompletion` independently.

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs (hunk 3 above)
Comment (davidfowl) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs]: It's extremely unlikely.

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs (hunk 5 above)
Comment (davidfowl) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs]: Use a single object field? I'm going to look at reducing the size of the pipe, it's one of the bigger objects we allocate today...

Comment (pakrym) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs]: Callback is action and a state and `PipeCompletionCallback ` is a struct.

**Pull request merged**